pkg/apis/configuration/v1alpha1/kongcustomentity_types.go

```diff
@@ -0,0 +1,76 @@
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// KongCustomEntityList is a list of kong custom entities.
+type KongCustomEntityList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []KongCustomEntity `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kongcustomentities,shortName=kce,categories=kong-ingress-controller
+
+// KongCustomEntity represents a custom entity in Kong.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Type is the type of this custom entity.
+	// Should be same as the `Name` of a KongCustomEntityDefinition.
+	Type string `json:"type"`
+	// Fields is the list of fields in the entity.
+	Fields apiextensionsv1.JSON `json:"fields"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=kongcustomentitydefinitions,shortName=kced,categories=kong-ingress-controller
+
+// KongCustomEntityDefinition represents definition of a custom entity type in Kong.
+type KongCustomEntityDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// Name is the type name of the entity.
+	Name string `json:"name"`
+	// AdminAPIName is the name used in admin API paths to CRUD this type of entity.
+	// If AdminAPIName is empty, it is seen as same as `Name`.
+	// For example: Name = "jwt_credentials" and AdminAPIName = "jwts", then we call `/jwts` or `/jwts/<id>` for CRUD.
+	AdminAPIName string `json:"adminAPIName,omitempty"`
+	// Dependecies are the entity types which are required by this type.
+	// If it is empty, the entity type is a "top level" object that does not dependent on other entities.
+	Dependecies []KongEntityForeignKey `json:"dependencies,omitempty"`
+}
+
+// +k8s:deepcopy-gen:
+
+// KongEntityForeignKey represents a foreign key constraint of Kong entity.
+type KongEntityForeignKey struct {
+	// +kubebuilder:validation:Required
+	// Type is the type of the dependent entity in the foreign key constraint.
+	Type string `json:"type"`
```

Comment:
Do you think we could make this validated enum so that's clear from an API standpoint what entity types we're expecting? Or will we allow anything here and validate only when building an actual Kong config?

As I understand, we cannot know what types might be expected here upfront because the relation might be to a custom entity which we do not know at this point because it's dynamically defined.

Reply:
Yeah, thanks for the answer, wanted to confirm my thinking is right 👍